### PR TITLE
Angelez/bibentries

### PR DESCRIPTION
**Pull request by geli-gel**
This is my PR for https://github.com/allenai/scholar/issues/33041 step 1: move code to mmda

In moving the code over I learned I had to figure out how the outputs will be used in SPP / stored in the annotation store, and talked w/Kyle and realized the importance of including the raw model output, which is why a `Prediction` includes `bib_entries` which contain boxes tightened around the tokens found inside of them + text and spans, as well as `original_boxes` which are raw model output detected boxes.

There are optional config variables: `BIB_ENTRY_DETECTION_PREDICTOR_SCORE_THRESHOLD` (Prediction accuracy score used to determine threshold of returned predictions) and `BIB_ENTRY_DETECTION_MIN_VILA_BIB_ROWS` (Minimum number of rows in a Bibliography VILA SpanGroup required to qualify as a Bibliography section) that can be adjusted if e2e eval shows signals we should change them.

### CODE REVIEW

File: mmda/predictors/d2_predictors/bibentry_detection_predictor.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+        page_width, page_height = image.size
+
+        return [
+            SpanGroup(
+                spans=[
+                    Span(
+                        # TODO: these start/ends are page token IDs?
+                        start=,
+                        end=,
```
Comment (geli-gel) [inline mmda/predictors/d2_predictors/bibentry_detection_predictor.py:66]: @cmwilhelm This is question #1

File: mmda/predictors/d2_predictors/bibentry_detection_predictor.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            List[BoxGroup]:
+                The returned Boxgroups for the detected layouts for all pages
+        """
+        document_prediction = []
+
+        for image_index, image in enumerate(tqdm(document.images)):
+            model_outputs = self.model.detect(image)
+            document_prediction.extend(
+                # TODO tokens .. ?
+                self.postprocess(model_outputs, document.tokens[image_index], image_index, image)
```
Comment (geli-gel) [inline mmda/predictors/d2_predictors/bibentry_detection_predictor.py:102]: @cmwilhelm This is question #2 - how do I learn more about how tokens exist in the Document that will be given as input? I'm assuming I can make it a requirement that the Document / Instance input contains Tokens per page somehow, but I'm struggling to understand how to match those tokens to what my bounding boxes contain. In my other code I use layoutparser's Layout.filter_by: `bib_tokens = page_tokens.filter_by(bib_block, center=True)`

**Review by cmwilhelm — CHANGES_REQUESTED** [review]
TIMO interface looks good to me. I'll defer to @kyleclo on the particulars of the mmda predictor since there's a fair amount of logic on using the vila info heuristically.

Just a couple requested changes:

1. Let's give `original_boxes` a more specific name since it's at the top-level document namespace.
2. We probably shouldn't depend on `vila_predictors` for your model. From what I can tell that's just there to allow the integration test to process the PDF before handing to your new predictor. I think we want to do that offline and generate a fixture JSON file instead.

File: ai2_internal/config.yaml [hunk 3]
```diff
@@ -167,3 +167,49 @@ model_variants:
...
+    # those in your setup.py.
+    docker_run_commands: [ "apt-get update && apt-get install -y poppler-utils libgl1",
+                           "pip install layoutparser",
+                           "pip install torch==1.8.0",
+                           "pip install torchvision==0.9.0",
+                           "pip install 'detectron2@git+https://github.com/facebookresearch/detectron2.git@v0.6#egg=detectron2'"]
+
+    # Any additional sets of dependencies required by the model.
+    # These are the 'extras_require' keys in your setup.py.
+    extras_require: [ "bibentry_detection_predictor", "vila_predictors" ]
```
[the remark this answers is not in the document]
Comment (geli-gel) [reply ai2_internal/config.yaml:215]: yes you're right - I'll change the test as you suggested!

File: ai2_internal/config.yaml [hunk 4]
```diff
@@ -167,3 +167,49 @@ model_variants:
...
+    predictor_config: ai2_internal.bibentry_detection_predictor.interface.PredictorConfig
+
+    # Full S3 path to tar.gz'ed artifacts archive, nullable
+    artifacts_s3_path: s3://ai2-s2-mmda/models/bibliography-entries/outputs/p3_2xl/bibs/silver_dataset_8k/one_category/pln_mask_rcnn_X_101_32x8d_FPN_3x/archive.tar.gz
+
+    # Version of python required for model runtime, e.g. 3.7, 3.8, 3.9
+    python_version: "3.8"
+
+    # Whether this model supports CUDA GPU acceleration
+    cuda: false
```
Comment (geli-gel) [inline ai2_internal/config.yaml:192]: whoops

File: ai2_internal/bibentries_v1/integration_test.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
+"""
```
[the remark this answers is not in the document]
Comment (geli-gel) [reply ai2_internal/bibentries_v1/integration_test.py:1]: yep, that's how it got here, my bad!

**Comment by geli-gel** [comment]
Going to do some changes to the output Prediction after the sync today - s2airs doesn't need the text stored here anymore, so just going to save boxes only. (output will be bib_entries: List[BoxGroup] and original_boxes: List[BoxGroup]

File: ai2_internal/bibentry_detection_predictor/integration_test.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+            symbols=doc.symbols,
+            tokens=tokens,
+            rows=rows,
+            pages=pages,
+            page_images=encoded_page_images,
+            vila_span_groups=vila_span_groups)]
+
+        predictions = container.predict_batch(instances)
+
+        for bib_entry in predictions[0].bib_entries:
```
[the remark this answers is not in the document]
Comment (geli-gel) [reply ai2_internal/bibentry_detection_predictor/integration_test.py:90]: I thought about making it more specific and decided against it but I think at least checking that the number of bib_entries and raw_bib_entry_boxes is the same would be a good addition

**Review by cmwilhelm — APPROVED** [review]
Thanks for the back and forth Angele. LGTM to merge after the merge conflict is resolved.

**Pull request merged**